module/zfs/zfs_vnops.c

```diff
@@ -3686,6 +3706,8 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 				VERIFY3U(zfs_link_destroy(tdl, szp, tx,
 				    ZRENAMING, NULL), ==, 0);
 			}
+		} else {
+			VERIFY0(zfs_link_create(tdl, tzp, tx, 0));
```

Comment:
Why is this only needed now?  This looks like it may address a potential existing issue.

Reply:
You are right, this seems like an existing issue. But, it is mostly a remote/unlikely case for the following reasons :
- We are deleting an entry on line:3677 and adding the same entry back on line:3680.
   Hence, i don't think it would fail.  Also, since this thread holds, the zfs_dirlock(), no other
   thread can race us.

However, as a safety measure, decided to add this code here.

Reply:
@sanjeevbagewadi 
I'll copy me gerrit comment here:
> You haven't add 'if (tzp)'
>
>Also, the zfs_link_destroy above will set z_unlinked to true, which will cause this to fail. And it will call zfs_unlinked_add. You'll need to revert those changes.

I don't feel we need to solve this in this patch, since this path is probably not very likely, and it's an existing issue. Maybe we should just have a warning here, and solve it later.

Reply:
Agreed. I had overlooked the fact the zfs_link_create() will not solve the problem. For now I will add an ASSERT() here. And we can take that up separately.